src/parse.rs

```diff
@@ -0,0 +1,566 @@
+                wasmparser::TypeDef::Instance(inst_ty) => {
+                    for e in inst_ty.exports.iter() {
+                        check_import_type(&types, is_root, &e.ty)?;
+                    }
+                    Ok(())
+                }
+                _ => unreachable!(),
+            }
+        }
+        wasmparser::ImportSectionEntryType::Memory(mem_ty) => match mem_ty {
+            wasmparser::MemoryType::M32 { limits: _, shared } => {
+                anyhow::ensure!(!shared, "shared memories are not supported by Wizer yet");
+                anyhow::ensure!(
+                    !is_root,
+                    "memory imports are not allowed in the root Wasm module"
+                );
+                Ok(())
+            }
+            wasmparser::MemoryType::M64 { .. } => {
+                anyhow::bail!("the memory64 proposal is not supported by Wizer yet")
+            }
+        },
+        wasmparser::ImportSectionEntryType::Event(_) => {
+            unreachable!("validation should have rejected the exceptions proposal")
+        }
+        wasmparser::ImportSectionEntryType::Function(_) => Ok(()),
+        wasmparser::ImportSectionEntryType::Table(_)
+        | wasmparser::ImportSectionEntryType::Global(_) => {
+            anyhow::ensure!(
+                !is_root,
+                "table and global imports are not allowed in the root Wasm module"
+            );
+            Ok(())
+        }
+    }
+}
+
+fn alias_section<'a>(
+    stack: &mut Vec<ModuleInfo<'a>>,
+    full_wasm: &'a [u8],
+    mut aliases: wasmparser::AliasSectionReader<'a>,
+) -> anyhow::Result<()> {
+    stack
+        .top_mut()
+        .add_raw_section(SectionId::Alias, aliases.range(), full_wasm);
+
+    // Clone any aliases over into this module's index spaces.
+    for _ in 0..aliases.get_count() {
+        let alias = aliases.read()?;
+        match &alias {
+            wasmparser::Alias::OuterType {
+                relative_depth,
+                index,
+            } => {
+                let relative_depth = usize::try_from(*relative_depth).unwrap();
+                let index = usize::try_from(*index).unwrap();
+                // NB: `- 2` rather than `- 1` because
+                // `relative_depth=0` means this module's immediate
+                // parent, not this module itself.
+                let ty = stack[stack.len() - 2 - relative_depth].types[index].clone();
+                stack.top_mut().types.push(ty);
+            }
+            wasmparser::Alias::OuterModule {
+                relative_depth,
+                index,
+            } => {
+                let relative_depth = usize::try_from(*relative_depth).unwrap();
+                let index = usize::try_from(*index).unwrap();
+                // Ditto regarding `- 2`.
+                let module = stack[stack.len() - 2 - relative_depth].modules[index].clone();
+                stack.top_mut().modules.push(module);
```

Comment:
Yeah, I was just kinda like yolo we can deal with it when its a problem, since I didn't want to deal with `Rc<RefCell<ModuleInfo>>` or putting everything into an arena.

Reply:
Latest commit addresses this (and fixes outer module alias support).